# Pertemuan 2| Sample Dataset pada Scikit-Learn | Belajar Machine Learning Dasar

### Load Dataset

Cara untuk mengakses dataset:
Kita dapat mengakses sebuah dataset dengan mengimport dataset terebut.

In [3]:
from sklearn.datasets import load_iris
#nama functionnya untuk melakukan load dataset 
#dataset yang di load adalah dataset iris

iris = load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [4]:
iris.keys()
#untuk mendata dictionary apa saja yang dimiliki oleh iris dataset

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

Isi keys dictionary dari iris dataset:

- data = akan berasosiasi dengan data-data features
- target = akan berasosiasi dengan data-data target
- frame = untuk menandakan apakah data formatnya merupakan data frame atau bukan. By default data formatnya adalah data array.
- target_name, features_name = akan berkorelasi dengan nama / label untuk setiap features atau targetnya
- DESCR = merupakan kependekan dari description, untuk mengakses deskripsi dari datasetnya.
- filename = akan berasosiasi dengan lokasi dari file dataset yang di load

### Deskripsi dari Sample Dataset

In [5]:
print(iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

Dalam deskripsi dari iris dataset dapat diketahui:
- Class yang terdapat pada deskripsi akan menjadi target
- Atribute yang terdapat pada deskripsi akan menjadi features

### Explanatory & Respone Variable (Feature & Target)

#### Explanatory Variables (Features)

In [6]:
X = iris.data
# untuk mengakses features yang akan ditampung didalam variable X
X.shape
# untuk mengakses dimensi dari datanya

X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

Penjelasan output: Dimensi datanya terdiri dari 150 baris dan 4 kolom.
X besar yang di comment dalam code digunakan untuk melihat datanya

#### Response Variable (Target)

In [7]:
y = iris.target
# untuk memanggil target yang akan dimasukkan kedalam variable y 
# y.shape
# untuk mengakses dimensi dari datanya
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Penjelasan output: Artinya data ini memiliki 150 baris dan 1 baris saja.
* Jumlah baris untuk features dan targetnya harus sama.
Jadi 1 set dari 1 baris atau 1 set features akan berkorelasi dengan 1 nilai di target.

### Features & Target Names

In [8]:
feature_names = iris.feature_names
# untuk memahami makna dari nilai-nilai yang ada di dalam array dataset sebelumnya
# untuk mengakses features name
feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

Urutan sangatlah penting, karena urutan akan merepresentasikan kolom yang ada pada dataset.

In [9]:
target_names = iris.target_names
# untuk memahami makna dari nilai-nilai yang ada di dalam array dataset sebelumnya
# untuk mengakses target name
target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

Target name akan menampilkan isi atau hasil dari dataset tersebut. 
Jadi pada dataset target sebelumnya dapat diambil kesimpulan bahwa:
    - 0 merepresentasikan sentosa
    - 1 merepresentasikan versicolor
    - 2 merepresentasikan virginica

### Visualisasi Data

#### Visualisasi Sepal Length & Width

In [10]:
import matplotlib.pyplot as plt
# untuk mengimport matplotlib sebagai plt

X = X[:, :2]
# artinya kita hanya menyertakan 2 kolom pertama saja dari X

x_min, x_max = X[:, 0].min() - 0.5, X[:, 0].max() + 0.5
# akan berasosiasi dengan nilai terkecil dan terbesar dari index ke 0 
# / berasosiasi dengna kolom sepal length
y_min, y_max = X[:, 1].min() - 0.5, X[:, 1].max() + 0.5
# akan berasosiasi dengan nilai terkecil dan terbesar dari index ke 0 
# / berasosiasi dengna kolom sepal width

# dan untuk nilai max dan minnya akan diberikan extra space sebesar 0.5 agar batasannya tidak terlalu ketat

plt.scatter(X[:, 0], X[:, 1], c=y)
# untuk membentuk scatter plot, sumbu x dari features index ke 0, sumbu y dari features index ke 1
# sumbu X = sepal lenght, sumbu y = sepal width
# c = color dan akan berasosiasi dengan nilai targetnya yang merepresentasikan nilai class atau spesiesnya

plt.xlabel('Sepal length')
plt.ylabel('Sepal width')

plt.xlim(x_min, x_max)
plt.ylim(y_min, y_max)
plt.grid(True)
plt.show()

<Figure size 640x480 with 1 Axes>

### Training & Testing Dataset

In [11]:
from sklearn.model_selection import train_test_split
# untuk membagi dataset menjadi training set dan testing set

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)
# untuk melakukan splitting dataset, proses spilitting ini dilakukan secara random

print(f'X train: {X_train.shape}')
print(f'X test: {X_test.shape}')
print(f'y train: {y_train.shape}')
print(f'y test: {y_test.shape}')
# untuk menampilkan shape / dimensi untuk setiap datasetnya

X train: (105, 2)
X test: (45, 2)
y train: (105,)
y test: (45,)


Penjelasan fungsi train_test_split:

* membutuhkan 4 parameter yaitu:
    * parameter 1 = features
    * parameter 2 = target
    * parameter 3 = test size (ukuran dari testing setnya, ukurannya di set dari 0.3 dalam presentase artinya test set nya adalah 30% sedangkan train set nya adalah 70%)
    * parameter 4 = untuk menseting random set number agar proses randomnya konsisten supaya bisa dilakukan replikasi
* fungsi trin_test_split akan menghasilkan 4 kelompok data (sehingga perlu ditampung dalam 4 variable) yaitu:
    * X_train = menampung features untuk training set
    * X_test = menampung features untuk testing set
    * y_train = menampung target untuk training set
    * y_test = menampung target untuk testing set

Penjelasan Output:
Memiliki jumlah barisnya 105 untuk training dataset baik X maupun Y, dan untuktesting set memiliki jumlah baris 45 baik untuk X maupun Y.

### Load Iris Dataset sebagai Pandas DataFrame

In [12]:
iris = load_iris(as_frame=True)
# kita load iris dataset dengan menyertakan parameter tambahan as_frame yang bernilai True
# untuk melakukan load iris dataset dengan pandas DataFrame
iris_features_df = iris.data
# untuk mengambil nilai featuresnya dan ditampung dalam variable iris_fratures_df
iris_features_df 

TypeError: load_iris() got an unexpected keyword argument 'as_frame'

In [14]:
import pandas as pd
print(pd.__version__)

0.25.1
